File: 3/js/src/QuickSort.ts
function quickSort(list: number[]) {
	quickSortHelper(list, 0, list.length - 1);
}

function quickSortHelper(list: number[], first: number, last: number) {
	if (first > last) {
		return;
	}

	const pivotIdx = partition(list, first, last);
	quickSortHelper(list, first, pivotIdx - 1);
	quickSortHelper(list, pivotIdx + 1, last);
}

function partition(list: number[], first: number, last: number): number {
	const pivotIdx = first;
	first++;

	while (first <= last) {
		if (list[first] > list[pivotIdx] && list[last] < list[pivotIdx]) {
			swap(first, last, list);
		}

		if (list[first] <= list[pivotIdx]) {
			first++;
		}

		if (list[last] >= list[pivotIdx]) {
			last--;
		}
	}

	swap(last, pivotIdx, list);

	return last;
}

function swap(i: number, j: number, list: number[]) {
	[list[i], list[j]] = [list[j], list[i]];
}

(function main() {
	const list = [8, 5, 2, 9, 5, 6, 3];
	quickSort(list);
	console.log(list);
})();

File: 3/java/Quicksort.java
import java.util.Arrays;

class Quicksort {
	public static void sort(int start, int end, int arr[]) {
		if(start <= end) {
			int pivotIdx = partition(start, end, arr);	
			sort(start, pivotIdx-1, arr);
			sort(pivotIdx+1, end, arr);
		}
	}

	public static void sort(int arr[]) {
		sort(0, arr.length-1, arr);	
	}
	
	public static void swap(int i, int j, int arr[]) {
		int temp = arr[i];
		arr[i] = arr[j];
		arr[j] = temp;
	}

	public static int partition(int start, int end, int arr[]) {
		int pivotIdx = start;
		int left = start + 1;
		int right = end;

		while(left <= right) {
			if(arr[left] > arr[pivotIdx] && arr[right] < arr[pivotIdx]) {
				swap(left, right, arr);
			}		
			
			if(arr[left] <= arr[pivotIdx]) {
				left++;
			}
			
			if(arr[right] >= arr[pivotIdx]) {
				right--;
			}
		}	

		swap(pivotIdx, right, arr);
		return right;
	}

	public static void main(String args[]) {
		int arr[] = {8,5,2,9,5,6,3};
		sort(arr);
		System.out.println(Arrays.toString(arr));
	}
}
